crates/topos-p2p/src/behaviour/gossip.rs

```diff
@@ -158,49 +156,42 @@ impl NetworkBehaviour for Behaviour {
         if self.tick.poll_tick(cx).is_ready() {
             // Publish batch
             if !self.echo_queue.is_empty() {
-                let mut echos = Batch { data: Vec::new() };
+                let mut echos = Batch {
+                    messages: Vec::new(),
+                };
                 for _ in 0..self.batch_size {
-                    if let Some(data) = self.echo_queue.pop_front() {
-                        echos.data.push(data);
+                    if let Some(request) = self.echo_queue.pop_front() {
+                        echos.messages.push(request);
                     } else {
                         break;
                     }
                 }
 
-                debug!("Publishing {} echos", echos.data.len());
-                if let Ok(msg) = bincode::serialize::<Batch>(&echos) {
-                    P2P_GOSSIP_BATCH_SIZE.observe(echos.data.len() as f64);
-
-                    match self.gossipsub.publish(IdentTopic::new(TOPOS_ECHO), msg) {
-                        Ok(message_id) => debug!("Published echo {}", message_id),
-                        Err(error) => error!("Failed to publish echo: {}", error),
-                    }
-                } else {
-                    P2P_MESSAGE_SERIALIZE_FAILURE_TOTAL
-                        .with_label_values(&["echo"])
-                        .inc();
+                debug!("Publishing {} echos", echos.messages.len());
+                let msg = echos.encode_to_vec();
+                P2P_GOSSIP_BATCH_SIZE.observe(echos.messages.len() as f64);
+                match self.gossipsub.publish(IdentTopic::new(TOPOS_ECHO), msg) {
+                    Ok(message_id) => debug!("Published echo {}", message_id),
+                    Err(error) => error!("Failed to publish echo: {}", error),
                 }
             }
 
             if !self.ready_queue.is_empty() {
-                let mut readies = Batch { data: Vec::new() };
+                let mut readies = Batch {
+                    messages: Vec::new(),
+                };
                 for _ in 0..self.batch_size {
                     if let Some(data) = self.ready_queue.pop_front() {
-                        readies.data.push(data);
+                        readies.messages.push(data);
                     } else {
                         break;
                     }
                 }
 
-                debug!("Publishing {} readies", readies.data.len());
-                if let Ok(msg) = bincode::serialize::<Batch>(&readies) {
-                    P2P_GOSSIP_BATCH_SIZE.observe(readies.data.len() as f64);
-                    _ = self.gossipsub.publish(IdentTopic::new(TOPOS_READY), msg);
-                } else {
-                    P2P_MESSAGE_SERIALIZE_FAILURE_TOTAL
-                        .with_label_values(&["ready"])
-                        .inc();
-                }
+                debug!("Publishing {} readies", readies.messages.len());
+                let msg = readies.encode_to_vec();
```

Comment:
This allocates again, so we're doing this three times I think? Once to put the message on the `VecDeque` queue, then again to build a `Vec` for the `Batch` and here one last time. Is there a way we can rethink this to avoid the overhead?

Reply:
I didn't mean to block this PR here, but perhaps there's some pattern to be figured out here that we can apply more broadly? 

/cc @Freyskeyd Any thoughts?

Reply:
Well, simply protobuf Rust structure is different from serialized binary message of that sturcture. We can maybe optimize passing of protobuf structure prior to serialization

Reply:
@dvdplm We are just passing message data around, we do not clone it. For the batch we temporary put it to the hash map queue and move it out from there. I don' think we can optimize allocation more than that